core/primitives-core/src/profile.rs

```diff
@@ -1,15 +1,25 @@
 use std::fmt;
+use std::ops::Index;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
 
 use crate::config::{ActionCosts, ExtCosts};
 use crate::types::Gas;
 
-/// Data for total gas burnt (index 0), and then each external cost and action
-type DataArray = [u64; ProfileData::LEN];
-
 /// Profile of gas consumption.
-#[derive(Clone, PartialEq, Eq)]
+/// Vecs are used for forward compatible. should only append new costs and never remove old costs
+#[derive(Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
 pub struct ProfileData {
-    data: DataArray,
+    all_gas: u64,
+    ext_costs: Vec<u64>,
+    action_costs: Vec<u64>,
+}
```

Comment:
I don't think that's the best representation for serialization. What this code does is it aligns the stored data with the *current* shape of code we use to access it. The problem is, the code is mutable, while the data lives forever. Serialized representation should be designed relative to the actual data we are storing, not relative to the actual code. 

In this particular case, the serialized repl reflects the fact that we store `all_gas` separately, and that we separte concpets for execution & action costs. The first is clearly just an accident -- we shouldn't be storing `all_gas` in the rocksdb at all, because that is alerady stored spearaetly as a part of `ExecutionOutcome`. 

I also don't think that it makes sense to separate action and execution costs in the database schema -- in the future, we might add a new kind of cost, or we might merge action and execution costs into one enum. 


I suggest using `Vec<u8>` as the borsh serialization -- this representation directly reflects the essential data we are storing -- a bunch of counters.

Reply:
> In this particular case, the serialized repl reflects the fact that we store all_gas separately, and that we separte concpets for execution & action costs. The first is clearly just an accident -- we shouldn't be storing all_gas in the rocksdb at all, because that is alerady stored spearaetly as a part of ExecutionOutcome.

But ProfileData::wasm_gas() do need total gas to figure out the wasm gas?

Other points sounds good.

Reply:
I check again, ProfileData::wasm_gas is not that important, we can drop it too. explorer/frontend can figure out that number from action cost and ext cost.

Reply:
I dropped all_gas and methods that refer to it. And also make all costs in one field, does not separate action cost and ext cost

> I suggest using Vec<u8> as the borsh serialization -- this representation directly reflects the essential data we are storing -- a bunch of counters.

Tried, however this make `Index` trait for `ProfileData` doesn't work well. We cannot tell arbitrary action cost is where quickly. I suggest store serialized repl also as `BTreeMap<Cost, u64>`

Reply:
@matklad I keep 2), the minimum changed one. Would like to know if you have a better solution or thought on approach 1) and 3).